docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+
+HOCs should pass through props that are unrelated to its specific concern. Most HOCs contain a render method that looks something like this:
+
+```js
+render() {
+  // Filter out extra props that are specific to this HOC and shouldn't be
+  // passed through
+  const { extraProp, ...passThroughProps } = this.props;
+
+  // Inject props into the wrapped component. These are usually state values or
+  // instance methods.
+  const injectedProp = someStateOrInstanceMethod;
+
+  // Pass props to wrapped component
+  return (
+    <WrappedComponent
+      injectedProp={injectedProp}
+      {...passThroughProps}
+    />
+  );
+}
+```
+
+This convention helps ensure that HOCs are as flexible and reusable as possible.
+
+### Optional: Maximizing Composability
+
+Not all HOCs look the same. Sometimes they accept only a single argument, the wrapped component:
+
+```js
+const NavbarWithRouter = withRouter(Navbar);
+```
+
+Usually, HOCs accept additional arguments. In this example from Relay, a config object is used to specify a component's data dependencies:
+
+```js
+const CommentWithRelay = Relay.createContainer(Comment, config);
+```
+
+By far, the most common signature for HOCs looks like this:
+
+```js
+// React Redux's `connect`
+const ConnectedComment = connect(commentSelector, commentActions)(Comment);
+```
+
+*What?!* If you break it apart, it's easier to see what's going on.
+
+```js
+// connect is a function that returns another function
+const enhance = connect(commentListSelector, commentListActions);
+// The returned function is an HOC, which returns a component that is connected
+// to the Redux store
+const ConnectedComment = enhance(CommentList);
+```
+In other words, `connect` is a higher-order function that returns a higher-order component!
+
+This form may seem confusing or unnecessary, but it has a useful property. Single-argument HOCs like the one returned by the `connect` function have the signature `Component => Component`. Functions whose output type is the same as its input type are really easy to compose together.
+
+```js
+// Instead of doing this...
+const EnhancedComponent = connect(commentSelector)(withRouter(WrappedComponent))
+
+// ... you can use a function composition utility
+// compose(f, g, h) is the same as (...args) => f(g(h(...args)))
+const enhance = compose(
+  // These are both single-argument HOCs
+  connect(commentSelector),
+  withRouter
+)
+const EnhancedComponent = enhance(WrappedComponent)
+```
+
+(This same property also allows `connect` and other enhancer-style HOCs to be used as decorators, an experimental JavaScript proposal.)
+
+The `compose` utility function is provided by many third-party libraries including lodash (as [`lodash.flowRight`](https://lodash.com/docs/#flowRight)), [Redux](http://redux.js.org/docs/api/compose.html), and [Rambda](http://ramdajs.com/docs/#compose).
+
+### Optional: Wrap the Display Name for Easy Debugging
```

Comment:
This "optional" thing seems just as conventiony as the other conventions. For consistency I'd just say "Convention: X" in the title of all the conventions, and then you can also ditch the little intro-to-the-conventions-section